### ISSUE TITLE
Allow using Ubuntu Docker Snap

### ISSUE THREAD

**Issue by madalinignisca** [microsoft/vscode-remote-release#5865]
I noticed two closed issues on same topic, but all the comments did not provide me facts about why you don't want to support Docker from snaps.

One mention was that it has no access to HOME folder. Actually that is not true. Docker installed from snap is limited to the User's Home folder by default, and with a small config change can be allowed to the full `/`. So, that should be clear, that any bind mount from user's folder is working 100%.

Docker snap provides both docker and docker-compose executables, both available for current user. Also, the service is accessible like docker.io and official build, it's socket is in `/var/run/docker.sock`. U can either use docker and docker-compose and connect thorough the socket to it directly.

Please don't hardcode blocking it, and simply provide a warning that you don't privide any support for people that are stubborn to use it.

**Comment by chrmarti** [comment]
We do not check if you are using Docker installed as snap. Maybe the Docker snap has changed since we last looked into this. What is the behavior / error you see?

**Comment by madalinignisca** [comment]
I get a simple message that I need docker-compose.

Basically, docker and docker-compose are available from the command line, are in the path.

As you support external docker on another server, I guessed u do simply use the available binaries.

**Comment by chrmarti** [comment]
Yes, we just call `docker` and `docker-compose` as they are on the `PATH`. You can call these from VS Code's integrated terminal, I guess?

**Comment by madalinignisca** [comment]
Yes. I can call them from Code's integrated terminal.
I thought VSCode might get started and won't have the final path that includes snap bin path and tried to set them also manually in `remote.containers.dockerComposePath` and `remote.containers.dockerPath`.

I don't know how it works for "classic" snaps (I thought that they work like common packages installed having full access everywhere), but on normal snaps, you'd need to make connection between a snap and another one and that grant's access to the binaries as well.

Why I find this good to fix as one use case:
Imagine a large company, which deploys Ubuntu on developer's machines and makes it's own snapcraft store internally. The machines do not grant root access to their users, and only snap is enabled, limited to the company's store.

Now probably some will say that both Code and Docker's repos should be present, and I'd say this is how it's done in most companies that enforce such strictness, but I do belive that next step from Canonical is an equivalent distribution of Ubuntu similar to what Fedora Silverblue is, where you will have a locked base OS and everything else coming from snap store, which can be a self maintained one by the company.

This would allow snap Code to use snap Docker and work in a strictly and very secured environment. On how snap is designed, once this link is done between snaps, should not affect at all current state of them.

**Comment by madalinignisca** [comment]
PS. I think I did posted an issue on another Code issue in snap, about it's hardware acceleration (snap is not using WebGL with hardware acceleration), as the snap is not setting a connection to WebGL.

For snaps to be able to do some stuff, they need connections. A hint can be here: https://ubuntu.com/blog/security-corner-snap-interface-snap-connections

**Comment by chrmarti** [comment]
From what I understand, the Docker snap would need to offer the commands as interfaces. What do you get for `docker interfaces docker`?

**Comment by tepene** [comment]
Since I already use VSCode as snap I would love to get it working with the docker snap.

When I start a new devcontainer I get `unable to prepare context: path "/tmp/vsch" not found`. So I assume we need to allow the snap docker acces to this folder? Or move the tmp folder to the home dir. Which would also fix #2347

is that the only limiting factor for using snap docker? or are there other things blocking this feature?

**complete devcontainer bring up log**
```
c10f59933ab3: Pull complete 
[2022-05-18T09:05:05.337Z] Digest: sha256:21a12816fcadaa16dabb4ba0e8c358361d02ea062b1b89db8786eb67173489d0
[2022-05-18T09:05:05.360Z] Status: Downloaded newer image for mcr.microsoft.com/vscode/devcontainers/python:0-3.10
[2022-05-18T09:05:05.392Z]  ---> 53f071753966
Step 3/4 : ARG NODE_VERSION="none"
[2022-05-18T09:05:06.294Z]  ---> Running in 66a74d52d211
[2022-05-18T09:05:06.537Z] Removing intermediate container 66a74d52d211
 ---> dfc3f1c61c72
Step 4/4 : RUN if [ "${NODE_VERSION}" != "none" ]; then su vscode -c "umask 0002 && . /usr/local/share/nvm/nvm.sh && nvm install ${NODE_VERSION} 2>&1"; fi
[2022-05-18T09:05:06.621Z]  ---> Running in 4831a254cc0f
[2022-05-18T09:05:08.607Z] Removing intermediate container 4831a254cc0f
 ---> bc247ec3a3c6
Successfully built bc247ec3a3c6
[2022-05-18T09:05:08.622Z] Successfully tagged vsc-test-29fce7ed9d9f5e5860a95092dd05a252:latest
[2022-05-18T09:05:08.637Z] Stop (144568 ms): Run: docker build -f /home/lus/Code/test/.devcontainer/Dockerfile -t vsc-test-29fce7ed9d9f5e5860a95092dd05a252 --build-arg VARIANT=3.10 --build-arg NODE_VERSION=none /home/lus/Code/test
[2022-05-18T09:05:08.644Z] Start: Run: docker inspect --type image vsc-test-29fce7ed9d9f5e5860a95092dd05a252
[2022-05-18T09:05:09.358Z] Stop (714 ms): Run: docker inspect --type image vsc-test-29fce7ed9d9f5e5860a95092dd05a252
[2022-05-18T09:05:09.362Z] Start: Run: docker build -f /tmp/vsch/updateUID.Dockerfile-0.234.0 -t vsc-test-29fce7ed9d9f5e5860a95092dd05a252-uid --build-arg BASE_IMAGE=vsc-test-29fce7ed9d9f5e5860a95092dd05a252 --build-arg REMOTE_USER=vscode --build-arg NEW_UID=1000 --build-arg NEW_GID=1000 --build-arg IMAGE_USER=root /tmp/vsch
[2022-05-18T09:05:09.560Z] unable to prepare context: path "/tmp/vsch" not found
[2022-05-18T09:05:09.565Z] Stop (203 ms): Run: docker build -f /tmp/vsch/updateUID.Dockerfile-0.234.0 -t vsc-test-29fce7ed9d9f5e5860a95092dd05a252-uid --build-arg BASE_IMAGE=vsc-test-29fce7ed9d9f5e5860a95092dd05a252 --build-arg REMOTE_USER=vscode --build-arg NEW_UID=1000 --build-arg NEW_GID=1000 --build-arg IMAGE_USER=root /tmp/vsch
[2022-05-18T09:05:09.568Z] Error: Command failed: docker build -f /tmp/vsch/updateUID.Dockerfile-0.234.0 -t vsc-test-29fce7ed9d9f5e5860a95092dd05a252-uid --build-arg BASE_IMAGE=vsc-test-29fce7ed9d9f5e5860a95092dd05a252 --build-arg REMOTE_USER=vscode --build-arg NEW_UID=1000 --build-arg NEW_GID=1000 --build-arg IMAGE_USER=root /tmp/vsch
[2022-05-18T09:05:09.568Z]     at IR (/home/lus/.vscode/extensions/ms-vscode-remote.remote-containers-0.234.0/dist/spec-node/devContainersSpecCLI.js:221:995)
[2022-05-18T09:05:09.569Z]     at zy (/home/lus/.vscode/extensions/ms-vscode-remote.remote-containers-0.234.0/dist/spec-node/devContainersSpecCLI.js:221:933)
[2022-05-18T09:05:09.569Z]     at runMicrotasks (<anonymous>)
[2022-05-18T09:05:09.569Z]     at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2022-05-18T09:05:09.569Z]     at async $R (/home/lus/.vscode/extensions/ms-vscode-remote.remote-containers-0.234.0/dist/spec-node/devContainersSpecCLI.js:226:2007)
[2022-05-18T09:05:09.569Z]     at async Zy (/home/lus/.vscode/extensions/ms-vscode-remote.remote-containers-0.234.0/dist/spec-node/devContainersSpecCLI.js:226:3112)
[2022-05-18T09:05:09.570Z]     at async BR (/home/lus/.vscode/extensions/ms-vscode-remote.remote-containers-0.234.0/dist/spec-node/devContainersSpecCLI.js:226:12448)
[2022-05-18T09:05:09.570Z]     at async qR (/home/lus/.vscode/extensions/ms-vscode-remote.remote-containers-0.234.0/dist/spec-node/devContainersSpecCLI.js:226:12204)
[2022-05-18T09:05:09.583Z] Stop (145978 ms): Run: /var/lib/snapd/snap/code/96/usr/share/code/code /home/lus/.vscode/extensions/ms-vscode-remote.remote-containers-0.234.0/dist/spec-node/devContainersSpecCLI.js up --user-data-folder /home/lus/.config/Code/User/globalStorage/ms-vscode-remote.remote-containers/data --container-data-folder .vscode-server/data/Machine --container-system-data-folder /var/vscode-server --workspace-folder /home/lus/Code/test --workspace-mount-consistency cached --id-label vsch.local.folder=/home/lus/Code/test --id-label vsch.quality=stable --log-level debug --log-format json --config /home/lus/Code/test/.devcontainer/devcontainer.json --default-user-env-probe loginInteractiveShell --mount type=volume,source=vscode,target=/vscode,external=true --skip-post-create --update-remote-user-uid-default on --mount-workspace-git-root true
[2022-05-18T09:05:09.585Z] Exit code 1
[2022-05-18T09:05:09.595Z] Command failed: /var/lib/snapd/snap/code/96/usr/share/code/code /home/lus/.vscode/extensions/ms-vscode-remote.remote-containers-0.234.0/dist/spec-node/devContainersSpecCLI.js up --user-data-folder /home/lus/.config/Code/User/globalStorage/ms-vscode-remote.remote-containers/data --container-data-folder .vscode-server/data/Machine --container-system-data-folder /var/vscode-server --workspace-folder /home/lus/Code/test --workspace-mount-consistency cached --id-label vsch.local.folder=/home/lus/Code/test --id-label vsch.quality=stable --log-level debug --log-format json --config /home/lus/Code/test/.devcontainer/devcontainer.json --default-user-env-probe loginInteractiveShell --mount type=volume,source=vscode,target=/vscode,external=true --skip-post-create --update-remote-user-uid-default on --mount-workspace-git-root true
[2022-05-18T09:05:09.595Z] Exit code 1
```

**Comment by tepene** [comment]
Issue at docker-snap repository @https://github.com/docker-snap/docker-snap/issues/58

**Comment by VSCodeTriageBot** [comment]
We closed this issue because we [don't plan to address it](https://aka.ms/vscode-out-of-scope) in the foreseeable future. If you disagree and feel that this issue is crucial: we are happy to listen and to reconsider.

If you wonder what we are up to, please see our [roadmap](https://aka.ms/vscoderoadmap) and [issue reporting guidelines](https://aka.ms/vscodeissuereporting).

Thanks for your understanding, and happy coding!
